### PR TITLE
SOT-124: ZK-Rollups after EIP-4844

### PR DESCRIPTION
**Pull request by hduoc2003**
# Check List

- [x] Did you check your grammar / reader proof your document using ChatGPT?
- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] After merging this, please update your task status!
- [x] Did you verify that your math formulas are rendered correctly?

### CODE REVIEW

File: terms/barycentric_equation.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Barycentric Equation
+
+**References:**
+
+- [A quick barycentric evaluation tutorial](https://hackmd.io/@vbuterin/barycentric_evaluation)
+
+For a given set of pairs $(x_1;y_1), (x_2; y_2), ..., (x_N; y_N)$, let $P(x)$ be the polynomial such that $P(x_i) = y_i \ \forall
+i=1,...,N$. To evaluate $P(x)$ at an arbitrary point, we can interpolate $P(x)$
+using [Lagrange interpolation](lagrange_interpolation.md), but this technique allows us to evaluate $P(x)$ directly in $O(N)$ time. For
```
Comment (VanhGer) [inline terms/barycentric_equation.md:9]: rewrite this sentence

File: terms/barycentric_equation.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# Barycentric Equation
+
+**References:**
+
+- [A quick barycentric evaluation tutorial](https://hackmd.io/@vbuterin/barycentric_evaluation)
+
+For a given set of pairs $(x_1;y_1), (x_2; y_2), ..., (x_N; y_N)$, let $P(x)$ be the polynomial such that $P(x_i) = y_i \ \forall
+i=1,...,N$. Usually, to evaluate $P(x)$ at an arbitrary point, we can interpolate $P(x)$
+using [Lagrange interpolation](lagrange_interpolation.md) in $O(N^2)$ time in general case, and in $O(Nlog_2N)$ in special cases.
+Using barycentric formula, we can evaluate $P(x)$ directly in $O(N)$ time (with $O(n^{2})$ pre-computed time in general case). For
```
Comment (VanhGer) [inline terms/barycentric_equation.md:10]: n or N ?

Comment (VanhGer) [inline terms/barycentric_equation.md:10]: should add linking words like: However, But,..

**Review by VanhGer — APPROVED** [review]

**Review by ktokto313 — APPROVED** [review]

File: docs/zk_rollup_after_eip4844.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    assert kzg_to_versioned_hash(commitment) == versioned_hash
+
+    # Verify KZG proof with z and y in big endian format
+    assert verify_kzg_proof(commitment, z, y, proof)
+
+    # Return FIELD_ELEMENTS_PER_BLOB and BLS_MODULUS as padded 32 byte big endian values
+    return Bytes(U256(FIELD_ELEMENTS_PER_BLOB).to_be_bytes32() + U256(BLS_MODULUS).to_be_bytes32())
+```
+
+The `point_evaluation_precompile(versioned_hash, kzg_commitment, proof, z, y)` receives a versioned hash, the KZG commitment to the
```
Comment (zk-steve) [inline docs/zk_rollup_after_eip4844.md:63]: wrong param order point_evaluation_precompile(versioned_hash, kzg_commitment, proof, z, y)

File: terms/barycentric_equation.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+$P(0), P(1), ..., P(100)$ to compute $P(101)$, or $P(12042003)$, without ever reconstructing the polynomial.
+
+## General Technique
+
+From Lagrange interpolation, we have $P(x) = \sum_{i}{y_i L_i(x)}$. Let us denote:
+
+- $d_i=\prod_{j \neq i}{(x_i - x_j)}$
+- $M(x) = \prod_{i}{x-x_i}$
+
+We can now express $L_i(x)$ as $\dfrac{M(x)}{d_i(x-x_i)}$ for $(x \neq x_i)$. It takes $O(n^2)$ time to compute all $d_i$, $O(N)$ time
```
Comment (zk-steve) [inline terms/barycentric_equation.md:21]:
```suggestion
We can now express $L_i(x)$ as $\dfrac{M(x)}{d_i(x-x_i)}$ for $(x \neq x_i)$. It takes $O(N^2)$ time to compute all $d_i$, $O(N)$ time
```

File: terms/barycentric_equation.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+for $M(x)$ and $L_i(x)$. Hence, the expression:
+
+
+$$
+P(x) = M(x) * \sum_{i}{\frac{y_i}{d_i(x-x_i)}}
+$$
+
+
+can be computed in $O(N)$ time.
```
Comment (zk-steve) [inline terms/barycentric_equation.md:30]:
```suggestion
can be computed in $O(N)$ time if we cache the results of $d_i$
```

File: terms/barycentric_equation.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+**References:**
+
+- [A quick barycentric evaluation tutorial](https://hackmd.io/@vbuterin/barycentric_evaluation)
+
+For a given set of pairs $(x_1;y_1), (x_2; y_2), ..., (x_N; y_N)$, let $P(x)$ be the polynomial such that $P(x_i) = y_i \ \forall
+i=1,...,N$. Usually, to evaluate $P(x)$ at an arbitrary point, we can interpolate $P(x)$
+using [Lagrange interpolation](lagrange_interpolation.md) in $O(N^2)$ time in general case, and in $O(Nlog_2N)$ in special cases.
+Alternatively, utilizing the barycentric formula allows for direct evaluation of $P(x)$ in $O(N)$ time, with pre-computed time 
+complexity of $O(N^2)$ in the general case. For example, if $P(x)$ is a degree-100 polynomial, you can use the evaluations
```
Comment (zk-steve) [inline terms/barycentric_equation.md:11]: Linking problem. Rearrange sentences

File: docs/zk_rollup_after_eip4844.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+blob, and a KZG opening proof for point $z$ and value $y$ as input. It verifies that `kzg_commitment` corresponds to
+the `versioned_hash` provided and that the opening `proof` is valid. That is the for Lagrange interpolation polynomial of the
```
Comment (zk-steve) [inline docs/zk_rollup_after_eip4844.md:65]:
```suggestion
the `versioned_hash` provided and that the opening `proof` is valid. That is for the Lagrange interpolation polynomial of the
```

File: docs/zk_rollup_after_eip4844.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+blob `p(X)` we have `p(z) = y`.
+
+## Blob Consistency Check
+
+The main challenge we face is proving that $f$ indeed ==represents exactly the raw transaction data using circuits==. There is an
+approach in the case where the ZK rollup is BLS12-381 based, and a moderately easier approach for arbitrary ZK-SNARKs.
```
Comment (zk-steve) [inline docs/zk_rollup_after_eip4844.md:71]: .

File: docs/zk_rollup_after_eip4844.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+ZK rollup would provide two commitments to their transaction or state delta data: the blob commitment (which the protocol ensures
```
Comment (zk-steve) [inline docs/zk_rollup_after_eip4844.md:73]:
```suggestion
ZK rollups would provide two commitments to their transaction or state delta data: the blob commitment (which the protocol ensures
```

**Comment by zk-steve** [comment]
Need further info from these: 
https://github.com/mmjahanara/blob-consistency-check/blob/main/README.md
https://github.com/scroll-tech/scroll/blob/33089b829f72ce4ae56ba105248a1bbab9eb0d8e/contracts/src/L1/rollup/ScrollChain.sol#L164C25-L164C25
https://scroll.io/blog/data-availability-4844

**Comment by zk-steve** [comment]
The barycentric file is good to go but the rest would likely require significant changes. Split them into 2 PRs so we can merge the barycentric sooner.

File: docs/zk_rollup_after_eip4844.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    # Return FIELD_ELEMENTS_PER_BLOB and BLS_MODULUS as padded 32 byte big endian values
+    return Bytes(U256(FIELD_ELEMENTS_PER_BLOB).to_be_bytes32() + U256(BLS_MODULUS).to_be_bytes32())
+```
+
+The `point_evaluation_precompile(versioned_hash, kzg_commitment, proof, z, y)` receives a versioned hash, the KZG commitment to the
+blob, and a KZG opening proof for point $z$ and value $y$ as input. It verifies that `kzg_commitment` corresponds to
+the `versioned_hash` provided and that the opening `proof` is valid. That is the for Lagrange interpolation polynomial of the
+blob `p(X)` we have `p(z) = y`.
+
+## Blob Consistency Check
```
Comment (zk-steve) [inline docs/zk_rollup_after_eip4844.md:68]: How does this look like when implementing? Use the links provided in the general comment

Comment (zk-steve) [inline docs/zk_rollup_after_eip4844.md:68]: Some info about why we need to do this first

File: docs/zk_rollup_after_eip4844.md
Comment (zk-steve) [inline docs/zk_rollup_after_eip4844.md:1]: The structure of the sections is fragmented (hard to link them together)

File: presentations/dencun_upgrade_and_proto_danksharding.pptx
Comment (zk-steve) [inline presentations/dencun_upgrade_and_proto_danksharding.pptx:1]: Need a new presentation for this

File: docs/zk_rollups_after_eip4844.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+**References:**
+
+- [EIP-4844 Homepage](https://www.eip4844.com/#help-out)
+- [How to use KZG commitments in proofs](https://notes.ethereum.org/@dankrad/kzg_commitments_in_proofs)
+- [Data Availability Post 4844](https://scroll.io/blog/data-availability-4844)
+
+## Introduction
+
+**EIP-4844** (or Proto-Danksharding) is the backbone of Ethereum Dencun upgrade. This upgrade focuses on Layer 2 scalability by
+introducing blob transactions and blob data, storage costs and gas fees have been significantly reduced.
```
Comment (zk-steve) [inline docs/zk_rollups_after_eip4844.md:12]:
```suggestion
introducing blob transactions and blob data, resulting in storage costs and gas fees have been significantly reduced.
```

File: docs/zk_rollups_after_eip4844.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+The `point_evaluation_precompile(versioned_hash, z, y, kzg_commitment, proof)` receives a versioned hash, the KZG commitment to the
+blob, and a KZG opening proof for point $z$ and value $y$ as input. It verifies that `kzg_commitment` corresponds to
+the `versioned_hash` provided and that the opening `proof` is valid (i.e. $f(z)
+=y$).
+
+## Blob Consistency Check
+
+Since rollup contract only can access to versioned hash instead of the raw transaction data (which was previously included
+as `calldata`), the main challenge we face is proving that $f$ indeed ==represents exactly the raw transaction data using circuits==.
```
Comment (0x5ea000000) [inline docs/zk_rollups_after_eip4844.md:75]: space missing near "=="?

File: docs/zk_rollups_after_eip4844.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+4. Generate proof $\pi_2$, that proves $y = f(x)$ with respect to $C_2$
+
+The prover sends $C_1,C_2,y,\pi_1, \pi_2$. The verifier accepts if $y = f(hash(C_1,C_2))$ and the proofs $\pi_1$ and $\pi_2$ verify.
+This technique was summarized by
+Vitalik [here](https://ethresear.ch/t/easy-proof-of-equivalence-between-multiple-polynomial-commitment-schemes-to-the-same-data/8188).
+
+### With Any ZK-SNARKs
+
+We also choose $x$ as in previous section, and evaluate [barycentric equation](../terms/barycentric_equation.md) $f(x) = \dfrac{x^N}{N}
+\cdot \sum_{i}{\dfrac{d_i \cdot \omega^i}{x - \omega^i}}$ at our random $x$. This evaluation has to be done over BLS12-381 in our
```
Comment (0x5ea000000) [inline docs/zk_rollups_after_eip4844.md:98]: what is 'our circuit'?

File: docs/zk_rollups_after_eip4844.md [hunk 14]
```diff
@@ -152,41 +153,35 @@ err = ethclient.Client.SendTransaction(context, signedTx)
...
-    {
-        (bool success, bytes memory data) = POINT_EVALUATION_PRECOMPILE_ADDR.staticcall(
-            abi.encodePacked(_blobVersionedHash, blobKZGProof)
-        );
-        // We verify that the point evaluation precompile call was successful by testing the latter 32 bytes of the
-        // response is equal to BLS_MODULUS as defined in https://eips.ethereum.org/EIPS/eip-4844#point-evaluation-precompile
-        if (!success) revert ErrorCallPointEvaluationPrecompileFailed();
-        (, uint256 result) = abi.decode(data, (uint256, uint256));
-        if (result != BLS_MODULUS) revert ErrorUnexpectedPointEvaluationPrecompileOutput();
+    function verify(
```
Comment (hduoc2003) [inline docs/zk_rollups_after_eip4844.md:167]: align code indent

File: docs/zk_rollups_after_eip4844.md (hunk 13 above)
Comment (zk-steve) [reply docs/zk_rollups_after_eip4844.md:103]: our own circuit

File: docs/zk_rollups_after_eip4844.md (hunk 14 above)
Comment (zk-steve) [reply docs/zk_rollups_after_eip4844.md:167]: leave this as is as there is no global function level in solidity (meaning it has to be inside a class, thus leading to the indentation)

Comment (hduoc2003) [reply docs/zk_rollups_after_eip4844.md:167]: However, the comment and the function should have the same indentation.

File: docs/zk_rollups_after_eip4844.md (hunk 13 above)
Comment (0x5ea000000) [reply docs/zk_rollups_after_eip4844.md:103]: where is this "our own circuit"? ;v

Comment (zk-steve) [reply docs/zk_rollups_after_eip4844.md:103]: the circuit that we define on our own that suits our needs (arbitrarily)

File: docs/zk_rollups_after_eip4844.md (hunk 14 above)
Comment (zk-steve) [reply docs/zk_rollups_after_eip4844.md:167]: You can change it if you want but for me it looks fine

**Pull request merged**